tests/unit/bokeh/models/test_sources.py

```diff
@@ -785,6 +785,53 @@ def test_data_column_lengths(self) -> None:
         assert len(warns) == 1
         assert str(warns[0].message) == "ColumnDataSource's columns must be of the same length. Current lengths: ('a', 3), ('b', 2)"
 
+class TestDataTable:
+    def test_from_data_with_dataframe(self):
+        df = pd.DataFrame({'A': [1, 2, 3], 'B': [4, 5, 6]})
+        table = DataTable.from_data(df)
+        assert isinstance(table, DataTable)
+        assert set(table.source.data) - {"index"} == {'A', 'B'}
+
+    def test_from_data_with_dict(self):
+        data = {'A': [1, 2, 3], 'B': [4, 5, 6]}
+        table = DataTable.from_data(data)
+        assert isinstance(table, DataTable)
+        assert set(table.source.data) == {'A', 'B'}
+
+    def test_from_data_with_columndatasource(self):
+        data = ColumnDataSource({'A': [1, 2, 3], 'B': [4, 5, 6]})
+        table = DataTable.from_data(data)
+        assert isinstance(table, DataTable)
+        assert set(table.source.data) == {'A', 'B'}
+
+    def test_from_data_with_columns(self):
+        df = pd.DataFrame({'A': [1, 2, 3], 'B': [4, 5, 6], 'C': [7, 8, 9]})
+        table = DataTable.from_data(df, columns=['A', 'B'])
+        assert isinstance(table, DataTable)
+        assert set(table.source.data) == {'A', 'B'}
+
+        source = ColumnDataSource({'A': [1, 2, 3], 'B': [4, 5, 6], 'C': [7, 8, 9]})
+        table = DataTable.from_data(source, columns=['A', 'C'])
+        assert isinstance(table, DataTable)
+        assert set(table.source.data) == {'A', 'C'}
+
+    def test_from_data_with_dataclone(self):
+        source = ColumnDataSource({'A': [1, 2, 3], 'B': [4, 5, 6]})
+        table = DataTable.from_data(source)
+        assert source is not table.source
+
+        data = {'A': [1, 2, 3], 'B': [4, 5, 6]}
+        table = DataTable.from_data(data)
+        assert data is not table.source
+
+        df = pd.DataFrame({'A': [1, 2, 3], 'B': [4, 5, 6]})
+        table = DataTable.from_data(df)
+        assert df is not table.source
```

Comment:
```suggestion
    def test_from_data_with_dataclone(self):
        source = ColumnDataSource({'A': [1, 2, 3], 'B': [4, 5, 6]})
        table = DataTable.from_data(source)
        assert source is not table.source
```

I think the last two can be omitted, there is no way that they are the same object since the types are different (e.g. `dict` and CDS, or DataFrame and CDS) The case we want to protect against is modifying a user-supplied CDS

Reply:
Yeah, that makes sense.